File: JavaScript/leetcode75/src/main/Problem65_BestTimeToBuyAndSellStockWithTransactionFee.js
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */
var maxProfit = function (prices, fee) {
  //Edge case
  if (prices.length < 2) {
    return 0;
  }
  //chossen first share and added transaction fee
  let buyProfit = -prices[0] - fee;
  let sellProfit = 0;

  for (let price of prices) {
    //keeping track of last buy of share
    let lastBuy = buyProfit;
    //check to hold the share or buy
    buyProfit = Math.max(buyProfit, sellProfit - price - fee);
    //check if sell profit is more than previous, then sell and update the sellProfit
    sellProfit = Math.max(sellProfit, lastBuy + price);
  }
  return sellProfit;
};

module.exports = maxProfit;

File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem65_BestTimeToBuyAndSellStockWithTransactionFee.java
package org.phoenix.leetcode.challenges;

public class Problem65_BestTimeToBuyAndSellStockWithTransactionFee {
    public int maxProfit(int[] prices, int fee) {
        // Length of the array
        int n = prices.length;
        // Special case
        if (n < 2) {
            return 0;
        }
        // Price on which we buy the stock
        int buy = -prices[0] - fee;
        // Price on which we sell the stock
        int sell = 0;
        // Loop through the array
        for (int i = 1; i < n; i++) {
            // Last buying price
            int lastBuy = buy;
            buy = Math.max(buy, sell - prices[i] - fee);
            sell = Math.max(sell, lastBuy + prices[i]);
        }
        return sell;
    }
}
